ethereum/statetransition/src/main/java/tech/pegasys/teku/statetransition/datacolumns/CustodySync.java

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright Consensys Software Inc., 2024
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.statetransition.datacolumns;
+
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+import tech.pegasys.teku.ethereum.events.SlotEventsChannel;
+import tech.pegasys.teku.infrastructure.async.SafeFuture;
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
+import tech.pegasys.teku.spec.datastructures.blobs.versions.electra.DataColumnSidecar;
+
+public class CustodySync implements SlotEventsChannel {
+
+  private final DataColumnSidecarCustody custody;
+  private final DataColumnSidecarRetriever retriever;
+  private final int maxPendingColumnRequests = 1024;
+  private final int minPendingColumnRequests = 512;
+
+  private Map<ColumnSlotAndIdentifier, PendingRequest> pendingRequests = new HashMap<>();
+  private boolean started = false;
+
+  public CustodySync(DataColumnSidecarCustody custody, DataColumnSidecarRetriever retriever) {
+    this.custody = custody;
+    this.retriever = retriever;
+  }
+
+  private synchronized void onRequestComplete(PendingRequest request) {
+    DataColumnSidecar result = request.columnPromise.join();
+    custody.onNewValidatedDataColumnSidecar(result);
+    pendingRequests.remove(request.columnId.identifier());
+    fillUpIfNeeded();
+  }
+
+  private void fillUpIfNeeded() {
+    if (started && pendingRequests.size() <= minPendingColumnRequests) {
```

Comment:
not sure if I'm not wrong here, but say you started, fill up 800, next you got disconnection so noone hit `onRequestComplete`, and after that you will never enter `fillUp()` again

Reply:
We are filling up in `onSlot()` call as well.  
Just a note: Basically I'm expecting the `Retriever` would do its attempts infinitely until success or being explicitly cancelled()

Reply:
> We are filling up in onSlot() call as well.

yeah, but 800 > minPendingColumnRequests

> Basically I'm expecting the Retriever would do its attempts infinitely until success 

ahh got it, if so, all should be good